### PR TITLE
Read/Write Lock

### PR DESCRIPTION
**Pull request by ColdenCullen**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/rw_lock_test.c [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    ASSERT_SUCCESS(aws_thread_join(&thread), "Thread join failed with error code %d.", aws_last_error());
+    ASSERT_INT_EQUALS(rw_lock_data.thread_fn_increments, rw_lock_data.max_counts, "Thread 2 should have written all data");
+    ASSERT_INT_EQUALS(rw_lock_data.max_counts, rw_lock_data.counter, "Both threads should have written exactly the max counts.");
+
+    aws_thread_clean_up(&thread);
+    aws_rw_lock_clean_up(&rw_lock_data.rw_lock);
+
+    return 0;
+}
+AWS_TEST_CASE(rw_lock_is_actually_rw_lock_test, test_rw_lock_is_actually_rw_lock)
```
Comment (RandyGaul) [inline tests/rw_lock_test.c]: I’d add in a test to hammer the rw lock with lots of readers and a couple writers.

File: include/aws/common/rw_lock.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * you should not treat it like one. On platforms it is possible for it to be non-reentrant, it will be.
+*/
+AWS_COMMON_API int aws_rw_lock_try_rdlock(struct aws_rw_lock *lock);
+AWS_COMMON_API int aws_rw_lock_try_wrlock(struct aws_rw_lock *lock);
+
+/**
+ * Releases the lock.
+ */
+AWS_COMMON_API int aws_rw_lock_rdunlock(struct aws_rw_lock *lock);
+AWS_COMMON_API int aws_rw_lock_wrunlock(struct aws_rw_lock *lock);
```
[the remark this answers is not in the document]
Comment (ColdenCullen) [reply include/aws/common/rw_lock.h]: Sure, I was just trying to match pthread's api. But I think you're right, this'll be clearer.

**Comment by ColdenCullen** [comment]
Because the implementation is platform specific, and it seems silly to make these functions not inline-able. Re-install: they are excluded. You'll notice that `AWS_COMMON_OS_HEADERS` is populated per-platform.

File: include/aws/common/windows/rw_lock.inl [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+int aws_rw_lock_init(struct aws_rw_lock *lock) {
```
Comment (JonathanHenson) [inline include/aws/common/windows/rw_lock.inl]: what's the point of putting these in cpp guards if we aren't doing dll semantics?
Even then why would we export platform specific symbols?

Comment (ColdenCullen) [reply include/aws/common/windows/rw_lock.inl]: You're right, we probably don't need the cpp guards. Where do you see this exporting platform specific symbols?

File: include/aws/common/windows/rw_lock.inl [hunk 4]
```diff
@@ -0,0 +1,82 @@
+#ifndef AWS_COMMON_WINDOWS_RW_LOCK_INL
```
Comment (JonathanHenson) [inline include/aws/common/windows/rw_lock.inl]: I don't support of worrying about inlines here. At the point you're grabbing a lock, do you actually give a damn about a function call overhead?

Comment (ColdenCullen) [reply include/aws/common/windows/rw_lock.inl]: I was just doing what @graebm suggested. You guys fight it out.

File: include/aws/common/windows/rw_lock.inl (hunk 3 above)
Comment (graebm) [reply include/aws/common/windows/rw_lock.inl]: shouldn't these implementations be marked `inline` to prevent the compiler from borking?

**Pull request merged**